src/dag.jl

```diff
@@ -276,40 +276,49 @@ end
 ############################################################################
 
 #=
-    longestpath(adj) -> path
-Finds the critical path of a DAG G by using Dijsktra's shortest path algorithm on G
-Returns the nodes constituting the path
+    longest_path(graph) -> path
+
+Finds the critical path of a DAG `graph`, and returns the nodes constituting it
+in reverse order.
+
+Required interface for the graph:
+- topological_sort(graph) -> collection of nodes in `graph`, sorted in topological order
+- inneighbors(node)       -> collection of predecessors of `node` (identified by their tags)
+- weight(node)            -> cost of `node` (a positive Float64)
+- tag(node)               -> id of `node` (a strictly positive Int)
```

Comment:
> the tag method is also part of the required interface, and that it should never be 0 or the code will give the wrong answer.

Precisely! I tried specifying the broad lines of the interface in this comment. Does this answer your questions?

Since `Logger` is currently the only type implementing the interface, I was lazy and tried to re-use as many existing methods as possible. For example: `topological_sort` and `tag` already existed. `inneighbors` also existed (not as a method but as a field), and it identifies predecessors by their tags.

I decided to re-use all that in order to minimize the impact of the critical path algorithm on the code in `logger.jl`. An other option would have been to design the desired interface independently, and to adapt the `Logger` implementation to conform to it. It is not too late to adopt such a strategy.

Reply:
> Precisely! I tried specifying the broad lines of the interface in this comment. Does this answer your questions?

👍

Reply:
> Since `Logger` is currently the only type implementing the interface, I was lazy and tried to re-use as many existing methods as possible. For example: `topological_sort` and `tag` already existed. `inneighbors` also existed (not as a method but as a field), and it identifies predecessors by their tags.

Makes sense. The thing that got me confused is that the `DAG{DataFlowTasks}` has an `inneighbors` method that returns the nodes and not the tags. Also, it requires you to pass the `dag` as well since `DataFlowTask` does not store the neighbours (unlike the `TaskLog`) in its fields. Perhaps we should rename `inneighbors` to something like `intags` for the interface method to avoid possible confusion? Otherwise I am quite happy with the changes!

Reply:
> I decided to re-use all that in order to minimize the impact of the critical path algorithm on the code in `logger.jl`. An other option would have been to design the desired interface independently, and to adapt the `Logger` implementation to conform to it. It is not too late to adopt such a strategy.

I think your strategy is simple and works for what we need. In the future, should we need to compute the longest path for scheduling purposes on the flight, we can always add the interface you described to the `TaskGraph` type. Requiring a strictly positive (integer) tag is not something every dag would have I guess, but we are not building a graph package right? 🤣

Reply:
> Perhaps we should rename inneighbors to something like intags for the interface method to avoid possible confusion?

Nice idea. Done!

Reply:
> In the future, should we need to compute the longest path for scheduling purposes on the flight, we can always add the interface you described to the TaskGraph type. Requiring a strictly positive (integer) tag is not something every dag would have I guess, but we are not building a graph package right?

Yes, we can always implement the interface for the `TaskGraph` type later. But if/when that time comes, we can also change the interface in order for it to be easier to implement for `TaskGraph`. And it will be no big deal to adapt the implementation for `Logger` as well.

I considered adding functions like `tagtype(node)` and maybe re-use `zero(tagtype(node))` as the sentinel value for root nodes. This would make everything more generic, but it feels like it would be a bit over-engineered at this time.